nexus-repository-p2/src/main/java/org/sonatype/nexus/repository/p2/internal/proxy/P2ProxyFacetImpl.java

```diff
@@ -301,18 +303,31 @@ protected String getUrl(@Nonnull final Context context) {
   }
 
   @VisibleForTesting
-  protected P2Attributes mergeAttributesFromTempBlob(final TempBlob tempBlob, final P2Attributes p2Attributes) {
-    try (JarInputStream jis = getJar(tempBlob, p2Attributes.getExtension())) {
-      return jarParser
-          .getAttributesFromJarFile(jis)
-          .map(jarP2Attributes -> P2Attributes.builder().merge(p2Attributes, jarP2Attributes).build())
-          .orElse(p2Attributes);
+  protected P2Attributes mergeAttributesFromTempBlob(final TempBlob tempBlob, final P2Attributes sourceP2Attributes) {
+    checkNotNull(sourceP2Attributes.getExtension());
+
+    Optional<P2Attributes> p2Attributes = Optional.empty();
+    // first try Features XML
+    try (JarInputStream jis = getJar(tempBlob, sourceP2Attributes.getExtension())) {
+      p2Attributes = jarParser.getAttributesFromFeatureXML(jis);
+    }
+    catch (Exception ex) {
+      log.warn("Could not get attributes from feature.xml due to following exception: {}", ex.getMessage());
+    }
+
+    // second try Manifest
+    try (JarInputStream jis = getJar(tempBlob, sourceP2Attributes.getExtension())) {
+      if (!p2Attributes.isPresent()) {
```

Comment:
this condition can be out of opening the input stream

Reply:
Done